docs/dev/testing.rst

```diff
@@ -0,0 +1,306 @@
+locally, but that won't necessarily be installed by PUDL users. Tox's behavior
+is configured with the ``tox.ini`` file in the main repository directory. There
+are several different "test environments" defined, to test different aspects
+of the software, or to perform other actions like building the documentation.
+To see a list of the available environments with short descriptions in the PUDL
+repository run:
+
+.. code-block:: console
+
+    $ tox -av
+
+    default environments:
+    ci               -> Run all continuous integration (CI) checks & generate test coverage.
+
+    additional environments:
+    flake8           -> Run the full suite of flake8 linters on the PUDL codebase.
+    pre_commit       -> Run git pre-commit hooks not covered by the other linters.
+    bandit           -> Check the PUDL codebase for common insecure code patterns.
+    linters          -> Run the pre-commit, flake8 and bandit linters.
+    doc8             -> Check the documentation input files for syntactical correctness.
+    docs             -> Remove old docs output and rebuild HTML from scratch with Sphinx
+    unit             -> Run all the software unit tests.
+    ferc1_solo       -> Test whether FERC 1 can be loaded into the PUDL database alone.
+    integration      -> Run all software integration tests and process a full year of data.
+    validate         -> Run all data validation tests. This requires a complete PUDL DB.
+    ferc1_schema     -> Verify FERC Form 1 DB schema are compatible for all years.
+    full_integration -> Run ETL and integration tests for all years and data sources.
+    full             -> Run all CI checks, but for all years of data.
+    build            -> Prepare Python source and binary packages for release.
+    testrelease      -> Do a dry run of Python package release using the PyPI test server.
+    release          -> Release the PUDL package to the production PyPI server.
+
+Continuous Integration Tests
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Our default tox test environment is ``ci`` -- which includes all of the tests
+that will be run in continuous integration using a `GitHub Action
+<https://github.com/features/actions>`__. You should run these tests before
+pushing code to the repository or making a pull request. Because it's the
+default test environment, it will be run if you call Tox without any
+arguments:
+
+.. code-block:: console
+
+    $ tox
+
+This is equivalent to:
+
+.. code-block:: console
+
+    $ tox -e ci
+
+If the dependencies have been changed, or you recently ran the tests while on
+another branch of the repository with other dependencies, you should tell Tox
+to rebuild the software environment it operates in with the ``-r`` (aka
+``--recreate``) flag:
+
+.. code-block:: console
+
+    $ tox -re ci
+
+In addition to running the ``unit`` and ``integration`` tests, the CI test
+environment lints the code and documentation input files, and uses Sphinx
```

Comment:
maybe these should be their own testing category?

Reply:
What do you mean by "these" and "testing category?" The docs, linters, unit, and integration tests already exist as their own test environments (see the output from `tox -av` above). I was trying to avoid exhaustively documenting all of those different scenarios, and just focusing on the ones I think we probably want people to run regularly (`unit`, `integration`, `ci` and `validate`) which indirectly include pretty much all of the test environments. Are there other test environments within Tox that you think are worth highlighting in particular?

Reply:
the linters and documentation fall into what I think is the testing environment. it just feels semi-confusing to talk about the unit, integration, ci and validate tests and then the testing environment stuff as other more different tests.. when really these are integrated into a lot of these big test categories (unit, integration, ci and validate)?

Reply:
The linters and doc build are only run as part of the `ci` test environment, or if you specifically run them. `tox -e unit` only runs the pytest tests which are under `test/unit` and `tox -e integration` and `tox -e validate` similarly only run the tests that are defined under `test/integration` and `test/validate` respectively. I guess I'm trying to differentiate between the **tests** which are actual software tests, defined and run using `pytest` and... all of the other kinds of things which Tox is set up to do, including building docs, running linters, building the software package for distribution, uploading those packages to PyPI, running the `git pre-commit` hooks, etc.

Reply:
okay I think it would be important to give a liiiitle bit of background on that in the intro section to this page. the three categories in that section really got me wanting to see those three categories in all of the content below, which is sort of there, but just doesn't incorporate this kinda of thing.

it might be good to incorporate a "hey we use pytest and tox. breif description of difference. tox runs the pytest tests. the software tests are broken out into three categories: __, __, __. tox also tests the testing environment (?)"